semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala

```diff
@@ -0,0 +1,532 @@
+package scala.meta.internal.scalacp
+
+import java.nio.file._
+import scala.collection.mutable
+import scala.meta.internal.metacp._
+import scala.meta.internal.{semanticdb3 => s}
+import scala.meta.internal.semanticdb3.Accessibility.{Tag => a}
+import scala.meta.internal.semanticdb3.SymbolInformation.{Kind => k}
+import scala.meta.internal.semanticdb3.SymbolInformation.{Property => p}
+import scala.meta.internal.semanticdb3.SingletonType.{Tag => st}
+import scala.meta.internal.semanticdb3.Type.{Tag => t}
+import scala.reflect.NameTransformer
+import scala.tools.scalap._
+import scala.tools.scalap.scalax.rules.ScalaSigParserError
+import scala.tools.scalap.scalax.rules.scalasig._
+
+object Scalacp {
+  def parse(classfile: ToplevelClassfile): Option[ToplevelInfos] = {
+    // TODO: Parse scalaSig directly from classfile.node
+    // to avoid reading the bytes and parsing the classfile structure twice.
+    val bytes = Files.readAllBytes(classfile.path)
+    val scalapClassfile = ClassFileParser.parse(ByteCode(bytes))
+    ScalaSigParser.parse(scalapClassfile).map { scalaSig =>
+      val toplevels = scalaSig.topLevelClasses ++ scalaSig.topLevelObjects
+      val others = {
+        scalaSig.symbols.toList.flatMap {
+          case sym: SymbolInfoSymbol if !toplevels.contains(sym) => Some(sym)
+          case _ => None
+        }
+      }
+      val stoplevels = toplevels.flatMap(sinfo)
+      val sothers = toplevels.flatMap(spackages).distinct ++ others.flatMap(sinfo)
+      ToplevelInfos(classfile, stoplevels, sothers)
+    }
+  }
+
+  private def spackages(sym: SymbolInfoSymbol): List[s.SymbolInformation] = {
```

Comment:
Nitpick. I find it a bit weird to make as much as possible private in the internal package. These methods are not protected by any bincompat guarantees.

Reply:
This emphasizes that the methods aren't part of the interface between metacp, javacp and scalacp. However if you'd like, I can make them public again.

Reply:
I hit on myself the other day that I wanted to access a private method in Metap from external repo. In this particular case it's probably better to come up with a better public API for metap than un-private methods. Nothing to act upon now.